sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/faultinjection/SessionRetryOptionsTests.java

```diff
@@ -0,0 +1,437 @@
+        List<String> preferredLocations = this.writeRegionMap.keySet().stream().collect(Collectors.toList());
+        Duration sessionTokenMismatchDefaultWaitTime = Duration.ofMillis(Configs.getSessionTokenMismatchDefaultWaitTimeInMs());
+
+        assertThat(preferredLocations).isNotNull();
+        assertThat(preferredLocations.size()).isEqualTo(2);
+
+        CosmosAsyncClient clientWithPreferredRegions = new CosmosClientBuilder()
+            .endpoint(TestConfigurations.HOST)
+            .key(TestConfigurations.MASTER_KEY)
+            .consistencyLevel(BridgeInternal.getContextClient(this.cosmosAsyncClient).getConsistencyLevel())
+            .preferredRegions(preferredLocations)
+            .sessionRetryOptions(new CosmosSessionRetryOptionsBuilder()
+                .setRegionSwitchHint(regionSwitchHint)
+                .build()
+            )
+            .directMode()
+            .buildAsyncClient();
+
+        CosmosAsyncContainer containerForClientWithPreferredRegions = clientWithPreferredRegions
+            .getDatabase(this.cosmosAsyncContainer.getDatabase().getId())
+            .getContainer(this.cosmosAsyncContainer.getId());
+
+        TestItem createdItem = TestItem.createNewItem();
+        containerForClientWithPreferredRegions.createItem(createdItem).block();
+
+        FaultInjectionRuleBuilder badSessionTokenRuleBuilder = new FaultInjectionRuleBuilder("serverErrorRule-bad-session-token-" + UUID.randomUUID());
+
+        FaultInjectionCondition faultInjectionConditionForReadsInPrimaryRegion = new FaultInjectionConditionBuilder()
+            .operationType(faultInjectionOperationType)
+            .connectionType(FaultInjectionConnectionType.DIRECT)
+            .region(preferredLocations.get(0))
+            .build();
+
+        FaultInjectionServerErrorResult badSessionTokenServerErrorResult = FaultInjectionResultBuilders
+            .getResultBuilder(FaultInjectionServerErrorType.READ_SESSION_NOT_AVAILABLE)
+            .build();
+
+        FaultInjectionRule badSessionTokenRule = badSessionTokenRuleBuilder
+            .condition(faultInjectionConditionForReadsInPrimaryRegion)
+            .result(badSessionTokenServerErrorResult)
+            .duration(Duration.ofSeconds(10))
+            .build();
+
+        CosmosFaultInjectionHelper
+            .configureFaultInjectionRules(containerForClientWithPreferredRegions, Arrays.asList(badSessionTokenRule))
+            .block();
+
+        validateOperationExecutionResult(
+            performDocumentOperation(
+                containerForClientWithPreferredRegions,
+                createdItem,
+                operationType
+            ),
+            sessionTokenMismatchDefaultWaitTime,
+            regionSwitchHint);
+
+        safeCloseAsync(clientWithPreferredRegions);
```

Comment:
maybe put the close client logic in the finally block -> it will make sure the client always get a chance to be closed

Reply:
Fixed in next iteration.